sync-r2.sh

```diff
@@ -0,0 +1,99 @@
+cloudflare_cache_host="$5"
+cloudflare_api_token="$6"
+
+(( $# == 6 )) || die "Expected exactly 6 parameters."
+
+r2_endpoint="https://${r2_account_id}.r2.cloudflarestorage.com"
+r2_bucket_uri="s3://${r2_bucket_name}"
+
+echo "===> Using Cloudflare R2 bucket '${r2_bucket_name}' (zone: ${r2_zone}) via account ${r2_account_id}"
+
+[[ $(git rev-parse --is-shallow-repository) == false ]] || die "This script requires access to full git history to be able to set file timestamps correctly."
+
+echo "===> Updating file modification timestamps to match commits"
+# NOTE: `aws s3 sync` compares file timestamp and size to decide whether to upload it or not.
+readarray -t files < <(git ls-files)
+for file in "${files[@]}"; do
+    full_time="$(git log --max-count 1 --pretty=format:%cd --date=iso -- "$file")"
+    unix_timestamp="$(date --date="$full_time" +%Y%m%d%H%M.%S)"
+    touch -m -t "$unix_timestamp" "$file"
+done
+
+echo "===> Removing files that should not be uploaded to the bucket"
+# NOTE: This ensures that they will be deleted from the bucket if they're already there.
+# If we used `aws s3 sync --delete --exclude` instead, they would not get deleted.
+find . -path './.*' -delete
+find . -path './_*' -delete
+
+echo "===> Adding compatibility symlinks for files containing plus signs in the name"
+# NOTE: This is a quick'n'dirty workaround for Amazon S3 decoding plus sign in paths
+# as a space even though this substitution is only supposed to happen in a query string.
+# See https://forums.aws.amazon.com/thread.jspa?threadID=55746
+find . \
+    -regex "^\(.*/\)*[^/]*\+[^/]*$" \
+    -exec bash -c 'ln --symbolic --no-target-directory "$(basename "{}")" "$(dirname "{}")/$(basename "{}" | tr "+" " ")"' \;
+
+echo "===> Syncing binaries with the Cloudflare R2 bucket"
+aws --endpoint-url "$r2_endpoint" s3 sync . "$r2_bucket_uri" --delete --follow-symlinks --no-progress --exclude "*/list.*"
+
+echo "===> Syncing file lists with the Cloudflare R2 bucket"
+aws --endpoint-url "$r2_endpoint" s3 sync . "$r2_bucket_uri" --delete --follow-symlinks --no-progress --exclude "*" --include "*/list.*"
+
+echo "===> Purging Cloudflare CDN cache"
+# Purge only the files that might change in-place when new binaries are added.
+purge_paths=(
+    "/bin/soljson-nightly.js"
+    "/soljson.js"
+)
+while IFS= read -r path; do
+    purge_paths+=("/${path}")
+done < <(find . \( -wholename '*/list.*' -o -wholename '*/*-latest' -o -wholename '*/*-latest.*' \) | cut --characters 2-)
+
+purge_payload='{"files":['
+separator=""
+for path in "${purge_paths[@]}"; do
+    purge_payload+="${separator}\"https://${cloudflare_cache_host}${path}\""
+    separator=","
+done
+purge_payload+=']}'
+
+curl --fail --show-error --silent \
+    -X POST "https://api.cloudflare.com/client/v4/zones/${cloudflare_zone_id}/purge_cache" \
+    -H "Authorization: Bearer ${cloudflare_api_token}" \
+    -H "Content-Type: application/json" \
+    --data "$purge_payload" \
+    >/dev/null
```

Comment:
Not sure what the response looks like here, so maybe there is a good reason to hide it (very long, contains secrets, etc), but as a general rule I'd leave things like this in the output. Sometimes it helps figure out what's wrong when things do not work as expected.

```suggestion
curl --fail --show-error --silent \
    -X POST "https://api.cloudflare.com/client/v4/zones/${cloudflare_zone_id}/purge_cache" \
    -H "Authorization: Bearer ${cloudflare_api_token}" \
    -H "Content-Type: application/json" \
    --data "$purge_payload"
```